File: src/tools/terminal-exec.ts
// ============================================================================
// terminal_exec 工具——以“可执行文件 + 参数数组”的形式运行命令
// ============================================================================

import { spawn } from "node:child_process";
import { constants as fsConstants } from "node:fs";
import * as fs from "node:fs/promises";
import * as path from "node:path";
import { z } from "zod";
import type {
  ITool,
  ToolResult,
  ToolExecutionContext,
  JSONSchema7,
} from "./types.js";
import type { PermissionLevel } from "../core/types.js";
import {
  resolveExistingWorkspacePath,
  resolveWorkspaceDirectory,
  WorkspacePathError,
} from "./path-policy.js";

export interface TerminalExecInput {
  command: string;
  args?: string[];
  cwd?: string;
  timeoutMs?: number;
  env?: Record<string, string>;
}

export interface TerminalExecOutput {
  exitCode: number | null;
  signal: string | null;
  stdout: string;
  stderr: string;
  outputSummary: string;
  truncated: boolean;
}

const MAX_OUTPUT_LENGTH = 100_000;
const DEFAULT_TIMEOUT_MS = 120_000;
const COMMAND_PATTERN = /^[A-Za-z0-9._+-]+$/;

const terminalExecInputValidator = z.object({
  command: z.string().min(1).max(128).regex(
    COMMAND_PATTERN,
    "command 只能是单个可执行文件名；参数必须放入 args，不能包含空格、路径或 shell 元字符"
  ),
  args: z.array(z.string().max(4096)).max(128).optional(),
  cwd: z.string().min(1).max(4096).optional(),
  timeoutMs: z.number().int().min(1).max(DEFAULT_TIMEOUT_MS).optional(),
  env: z.record(z.string().max(32_768)).optional(),
}).strict();

type SafeCommandCheck = { allowed: true } | { allowed: false; reason: string };

function checkGitArgs(args: string[]): SafeCommandCheck {
  const subcommand = args[0];
  if (!subcommand || !["status", "diff", "log", "show"].includes(subcommand)) {
    return { allowed: false, reason: "auto-safe 只允许 git status、diff、log 和 show" };
  }

  const forbidden = [
    "-C",
    "-c",
    "--git-dir",
    "--work-tree",
    "--no-index",
    "--ext-diff",
    "--output",
    "--exec",
  ];
  if (args.some((arg) => forbidden.some((item) => arg === item || arg.startsWith(`${item}=`)))) {
    return { allowed: false, reason: "git 参数可能改变仓库边界、执行外部程序或写入文件" };
  }
  return { allowed: true };
}

async function checkLsArgs(args: string[], cwd: string): Promise<SafeCommandCheck> {
  for (const arg of args) {
    if (arg.startsWith("-")) continue;
    try {
      await resolveExistingWorkspacePath(arg, cwd);
    } catch {
      return { allowed: false, reason: `ls 路径不在工作区内或不存在: ${arg}` };
    }
  }
  return { allowed: true };
}

async function checkAutoSafeCommand(
  command: string,
  args: string[],
  cwd: string,
  env: Record<string, string> | undefined
): Promise<SafeCommandCheck> {
  if (env && Object.keys(env).length > 0) {
    return { allowed: false, reason: "auto-safe 不允许覆盖环境变量" };
  }
  if (command === "pwd") {
    return args.length === 0
      ? { allowed: true }
      : { allowed: false, reason: "pwd 不接受参数" };
  }
  if (command === "echo") return { allowed: true };
  if (command === "ls") return checkLsArgs(args, cwd);
  if (command === "git") return checkGitArgs(args);
  return { allowed: false, reason: `命令“${command}”不在 auto-safe 白名单中` };
}

function isWithinPath(root: string, target: string): boolean {
  const relative = path.relative(root, target);
  return relative === "" || (!relative.startsWith(`..${path.sep}`) && relative !== ".." && !path.isAbsolute(relative));
}

async function resolveSafeExecutable(
  command: string,
  workspace: string,
  env: Record<string, string>
): Promise<string | null> {
  const pathValue = env.PATH ?? env.Path ?? env.path ?? "";
  const extensions = process.platform === "win32"
    ? (env.PATHEXT ?? ".EXE;.CMD;.BAT").split(";")
    : [""];
  const realWorkspace = await fs.realpath(workspace);

  for (const pathEntry of pathValue.split(path.delimiter)) {
    if (!pathEntry || !path.isAbsolute(pathEntry)) continue;
    try {
      const realDirectory = await fs.realpath(pathEntry);
      if (isWithinPath(realWorkspace, realDirectory)) continue;
      for (const extension of extensions) {
        const candidate = path.join(realDirectory, `${command}${extension}`);
        try {
          const realCandidate = await fs.realpath(candidate);
          const stat = await fs.stat(realCandidate);
          if (!stat.isFile()) continue;
          if (process.platform !== "win32") {
            await fs.access(realCandidate, fsConstants.X_OK);
          }
          return realCandidate;
        } catch {
          // 当前 PATH 目录中不存在该可执行文件，继续查找。
        }
      }
    } catch {
      // 忽略不存在、不可访问或相对的 PATH 项。
    }
  }
  return null;
}

function failure(
  code: string,
  message: string,
  startedAt: string,
  details: Record<string, unknown> = {}
): ToolResult<TerminalExecOutput> {
  const endedAt = new Date().toISOString();
  return {
    success: false,
    output: null,
    error: { code, message, retryable: false, details },
    durationMs: new Date(endedAt).getTime() - new Date(startedAt).getTime(),
    startedAt,
    endedAt,
    metadata: {},
  };
}

export class TerminalExecTool implements ITool<TerminalExecInput, TerminalExecOutput> {
  name = "terminal_exec";
  description =
    "运行一个可执行文件并返回 stdout、stderr 和退出码。command 必须是单个可执行文件名，" +
    "所有参数必须放在 args 数组中。auto-safe 只允许 pwd、echo、ls 和只读 git 子命令；" +
    "其他命令需要 full-access。输出最多保留 100,000 个字符。";
  permissionLevel: PermissionLevel = "run-safe-commands";
  defaultTimeoutMs = DEFAULT_TIMEOUT_MS;
  inputValidator = terminalExecInputValidator;

  inputSchema: JSONSchema7 = {
    type: "object",
    properties: {
      command: {
        type: "string",
        pattern: COMMAND_PATTERN.source,
        description: "单个可执行文件名，例如 ls 或 git；不能包含空格和 shell 元字符",
      },
      args: {
        type: "array",
        items: { type: "string", maxLength: 4096 },
        description: "命令参数数组",
      },
      cwd: { type: "string", description: "工作区内的工作目录" },
      timeoutMs: {
        type: "number",
        minimum: 1,
        maximum: DEFAULT_TIMEOUT_MS,
        description: "超时时间，单位毫秒",
      },
      env: { type: "object", description: "仅 full-access 允许附加环境变量" },
    },
    required: ["command"],
    additionalProperties: false,
  };

  idempotencyKey(input: TerminalExecInput, context: ToolExecutionContext): string {
    return `terminal_exec:${input.command}:${JSON.stringify(input.args ?? [])}:${input.cwd ?? context.cwd}`;
  }

  async execute(
    input: TerminalExecInput,
    context: ToolExecutionContext
  ): Promise<ToolResult<TerminalExecOutput>> {
    const startedAt = new Date().toISOString();
    const args = input.args ?? [];

    let cwd: string;
    try {
      cwd = await resolveWorkspaceDirectory(input.cwd ?? context.cwd, context.cwd);
    } catch (error) {
      const message = error instanceof Error ? error.message : "工作目录无效";
      return failure(
        error instanceof WorkspacePathError ? error.code : "INVALID_CWD",
        message,
        startedAt,
        { cwd: input.cwd ?? context.cwd }
      );
    }

    let executable = input.command;
    if (context.permissions.mode !== "full-access") {
      const decision = await checkAutoSafeCommand(input.command, args, cwd, input.env);
      if (!decision.allowed) {
        return failure("UNSAFE_COMMAND", decision.reason, startedAt, {
          command: input.command,
          args,
          permissionMode: context.permissions.mode,
        });
      }
      const safeExecutable = await resolveSafeExecutable(input.command, context.cwd, context.env);
      if (!safeExecutable) {
        return failure(
          "SAFE_EXECUTABLE_NOT_FOUND",
          `无法在工作区外的可信 PATH 中找到命令“${input.command}”。`,
          startedAt,
          { command: input.command }
        );
      }
      executable = safeExecutable;
    }

    const timeoutMs = input.timeoutMs ?? this.defaultTimeoutMs;
    const childEnv = { ...context.env, ...(input.env ?? {}) };
    delete childEnv.GIT_EXTERNAL_DIFF;
    delete childEnv.GIT_CONFIG;
    delete childEnv.GIT_CONFIG_GLOBAL;
    delete childEnv.GIT_CONFIG_SYSTEM;

    return new Promise((resolve) => {
      let stdout = "";
      let stderr = "";
      let timedOut = false;
      let truncated = false;
      let settled = false;

      const finish = (result: ToolResult<TerminalExecOutput>) => {
        if (settled) return;
        settled = true;
        resolve(result);
      };

      const child = spawn(executable, args, {
        cwd,
        env: childEnv,
        shell: false,
        stdio: ["ignore", "pipe", "pipe"],
        signal: context.signal,
      });

      const timeout = setTimeout(() => {
        timedOut = true;
        child.kill("SIGTERM");
        const forceKill = setTimeout(() => child.kill("SIGKILL"), 3000);
        forceKill.unref();
      }, timeoutMs);

      child.stdout?.on("data", (chunk: Buffer) => {
        if (stdout.length >= MAX_OUTPUT_LENGTH) return;
        stdout += chunk.toString("utf-8");
        if (stdout.length >= MAX_OUTPUT_LENGTH) {
          truncated = true;
          stdout = stdout.slice(0, MAX_OUTPUT_LENGTH);
        }
      });

      child.stderr?.on("data", (chunk: Buffer) => {
        if (stderr.length >= MAX_OUTPUT_LENGTH) return;
        stderr += chunk.toString("utf-8");
        if (stderr.length >= MAX_OUTPUT_LENGTH) {
          truncated = true;
          stderr = stderr.slice(0, MAX_OUTPUT_LENGTH);
        }
      });

      child.on("error", (error) => {
        clearTimeout(timeout);
        finish(failure(
          context.signal.aborted ? "COMMAND_ABORTED" : "COMMAND_SPAWN_ERROR",
          context.signal.aborted ? "命令已中断" : error.message,
          startedAt,
          { command: input.command }
        ));
      });

      child.on("close", (exitCode, signal) => {
        clearTimeout(timeout);
        const endedAt = new Date().toISOString();
        const outputSummary = (stdout || stderr).split("\n").slice(-20).join("\n");
        finish({
          success: exitCode === 0 && !timedOut,
          output: {
            exitCode,
            signal,
            stdout,
            stderr,
            outputSummary,
            truncated: truncated || timedOut,
          },
          error: exitCode === 0 && !timedOut
            ? null
            : {
                code: timedOut ? "COMMAND_TIMEOUT" : "NONZERO_EXIT",
                message: timedOut
                  ? `命令在 ${timeoutMs}ms 后超时`
                  : `命令退出码为 ${exitCode}`,
                retryable: timedOut,
                details: { command: input.command, exitCode, signal, timeoutMs },
              },
          durationMs: new Date(endedAt).getTime() - new Date(startedAt).getTime(),
          startedAt,
          endedAt,
          metadata: { command: input.command, args, exitCode, signal, cwd: path.resolve(cwd) },
        });
      });
    });
  }
}

File: crates/xycli-core/src/tools/terminal_exec.rs
//! `terminal_exec`：不经过 shell，以“可执行文件 + 参数数组”执行命令。

use std::{
    collections::HashMap,
    path::{Path, PathBuf},
    process::Stdio,
    time::Duration,
};

use async_trait::async_trait;
use serde_json::{Value, json};
use tokio::{
    io::{AsyncRead, AsyncReadExt},
    process::Command,
    time,
};

use crate::permission::{PermissionLevel, PermissionMode};

use super::path_policy::{resolve_directory, resolve_existing};
use super::{
    Tool, ToolContext, ToolDefinition, ToolResult, object, reject_unknown_fields, required_string,
};

const MAX_OUTPUT_LENGTH: usize = 100_000;
const DEFAULT_TIMEOUT_MS: u64 = 120_000;

pub struct TerminalExecTool;

fn valid_command(command: &str) -> bool {
    !command.is_empty()
        && command.len() <= 128
        && command
            .bytes()
            .all(|byte| byte.is_ascii_alphanumeric() || b"._+-".contains(&byte))
}

fn check_git_args(args: &[String]) -> Result<(), String> {
    let Some(subcommand) = args.first() else {
        return Err("auto-safe 只允许 git status、diff、log 和 show。".into());
    };
    if !["status", "diff", "log", "show"].contains(&subcommand.as_str()) {
        return Err("auto-safe 只允许 git status、diff、log 和 show。".into());
    }
    let forbidden = [
        "-C",
        "-c",
        "--git-dir",
        "--work-tree",
        "--no-index",
        "--ext-diff",
        "--output",
        "--exec",
    ];
    if args.iter().any(|arg| {
        forbidden
            .iter()
            .any(|item| arg == item || arg.starts_with(&format!("{item}=")))
    }) {
        return Err("git 参数可能改变仓库边界、执行外部程序或写入文件。".into());
    }
    Ok(())
}

async fn check_auto_safe(
    command: &str,
    args: &[String],
    cwd: &Path,
    env: &HashMap<String, String>,
) -> Result<(), String> {
    if !env.is_empty() {
        return Err("auto-safe 不允许覆盖环境变量。".into());
    }
    match command {
        "pwd" if args.is_empty() => Ok(()),
        "pwd" => Err("pwd 不接受参数。".into()),
        "echo" => Ok(()),
        "ls" => {
            for arg in args.iter().filter(|arg| !arg.starts_with('-')) {
                resolve_existing(Path::new(arg), cwd)
                    .await
                    .map_err(|_| format!("ls 路径不在工作区内或不存在：{arg}"))?;
            }
            Ok(())
        }
        "git" => check_git_args(args),
        _ => Err(format!("命令“{command}”不在 auto-safe 白名单中。")),
    }
}

async fn resolve_safe_executable(command: &str, workspace: &Path) -> Option<PathBuf> {
    let real_workspace = tokio::fs::canonicalize(workspace).await.ok()?;
    let path = std::env::var_os("PATH")?;
    for directory in std::env::split_paths(&path) {
        if !directory.is_absolute() {
            continue;
        }
        let real_directory = match tokio::fs::canonicalize(&directory).await {
            Ok(value) => value,
            Err(_) => continue,
        };
        if real_directory.starts_with(&real_workspace) {
            continue;
        }
        let candidate = real_directory.join(command);
        let real_candidate = match tokio::fs::canonicalize(&candidate).await {
            Ok(value) => value,
            Err(_) => continue,
        };
        let metadata = match tokio::fs::metadata(&real_candidate).await {
            Ok(value) if value.is_file() => value,
            _ => continue,
        };
        #[cfg(unix)]
        {
            use std::os::unix::fs::PermissionsExt;
            if metadata.permissions().mode() & 0o111 == 0 {
                continue;
            }
        }
        return Some(real_candidate);
    }
    None
}

async fn read_limited(mut reader: impl AsyncRead + Unpin) -> (String, bool) {
    let mut retained = Vec::new();
    let mut buffer = [0_u8; 8192];
    let mut truncated = false;
    loop {
        let read = match reader.read(&mut buffer).await {
            Ok(0) | Err(_) => break,
            Ok(read) => read,
        };
        let remaining = MAX_OUTPUT_LENGTH.saturating_sub(retained.len());
        if remaining > 0 {
            retained.extend_from_slice(&buffer[..read.min(remaining)]);
        }
        if read > remaining {
            truncated = true;
        }
    }
    (String::from_utf8_lossy(&retained).into_owned(), truncated)
}

#[async_trait]
impl Tool for TerminalExecTool {
    fn definition(&self) -> ToolDefinition {
        ToolDefinition {
            name: "terminal_exec",
            description: "运行单个可执行文件并返回 stdout、stderr 和退出码。所有参数必须放入 args；auto-safe 仅允许 pwd、echo、ls 和只读 git 子命令。",
            input_schema: json!({
                "type":"object",
                "properties":{
                    "command":{"type":"string","pattern":"^[A-Za-z0-9._+-]+$"},
                    "args":{"type":"array","items":{"type":"string","maxLength":4096},"maxItems":128},
                    "cwd":{"type":"string","minLength":1,"maxLength":4096},
                    "timeoutMs":{"type":"integer","minimum":1,"maximum":DEFAULT_TIMEOUT_MS},
                    "env":{"type":"object","additionalProperties":{"type":"string","maxLength":32768}}
                },
                "required":["command"],
                "additionalProperties":false
            }),
            permission_level: PermissionLevel::RunSafeCommands,
            default_timeout: Duration::from_millis(DEFAULT_TIMEOUT_MS + 1_000),
        }
    }

    fn validate(&self, input: &Value) -> Result<(), Vec<String>> {
        let map = object(input)?;
        let mut issues = Vec::new();
        reject_unknown_fields(
            map,
            &["command", "args", "cwd", "timeoutMs", "env"],
            &mut issues,
        );
        match required_string(map, "command", 128, &mut issues) {
            Some(command) if !valid_command(command) => issues
                .push("command 只能是单个可执行文件名，不能包含路径、空格或 shell 元字符。".into()),
            _ => {}
        }
        if let Some(args) = map.get("args") {
            match args.as_array() {
                Some(args)
                    if args.len() <= 128
                        && args
                            .iter()
                            .all(|arg| arg.as_str().is_some_and(|value| value.len() <= 4096)) => {}
                _ => {
                    issues.push("args 必须是最多 128 项、单项不超过 4096 字节的字符串数组。".into())
                }
            }
        }
        if let Some(cwd) = map.get("cwd")
            && !cwd
                .as_str()
                .is_some_and(|value| !value.is_empty() && value.len() <= 4096)
        {
            issues.push("cwd 必须是 1 到 4096 字节的字符串。".into());
        }
        if let Some(timeout) = map.get("timeoutMs")
            && !timeout
                .as_u64()
                .is_some_and(|value| (1..=DEFAULT_TIMEOUT_MS).contains(&value))
        {
            issues.push(format!("timeoutMs 必须在 1 到 {DEFAULT_TIMEOUT_MS} 之间。"));
        }
        if let Some(env) = map.get("env") {
            match env.as_object() {
                Some(env)
                    if env
                        .values()
                        .all(|value| value.as_str().is_some_and(|value| value.len() <= 32_768)) => {
                }
                _ => issues.push("env 必须是字符串键值对象，值不超过 32768 字节。".into()),
            }
        }
        if issues.is_empty() {
            Ok(())
        } else {
            Err(issues)
        }
    }

    async fn execute(&self, input: Value, context: ToolContext) -> ToolResult {
        let command = input["command"].as_str().unwrap();
        let args = input
            .get("args")
            .and_then(Value::as_array)
            .map(|values| {
                values
                    .iter()
                    .map(|value| value.as_str().unwrap().to_owned())
                    .collect::<Vec<_>>()
            })
            .unwrap_or_default();
        let requested_cwd = input
            .get("cwd")
            .and_then(Value::as_str)
            .unwrap_or_else(|| context.cwd.to_str().unwrap_or("."));
        let cwd = match resolve_directory(Path::new(requested_cwd), &context.cwd).await {
            Ok(path) => path,
            Err(error) => {
                let code = if error.message.contains("超出工作区") {
                    "PATH_OUTSIDE_WORKSPACE"
                } else {
                    "INVALID_CWD"
                };
                return ToolResult::failure(
                    code,
                    error.message,
                    context.started_at,
                    json!({"cwd":requested_cwd}),
                );
            }
        };
        let env_overrides = input
            .get("env")
            .and_then(Value::as_object)
            .map(|env| {
                env.iter()
                    .map(|(key, value)| (key.clone(), value.as_str().unwrap().to_owned()))
                    .collect::<HashMap<_, _>>()
            })
            .unwrap_or_default();
        if context.permission_mode != PermissionMode::FullAccess {
            if let Err(reason) = check_auto_safe(command, &args, &cwd, &env_overrides).await {
                return ToolResult::failure(
                    "UNSAFE_COMMAND",
                    reason,
                    context.started_at,
                    json!({"command":command,"args":args,"permissionMode":context.permission_mode.as_str()}),
                );
            }
        }
        let executable = if context.permission_mode == PermissionMode::FullAccess {
            PathBuf::from(command)
        } else {
            match resolve_safe_executable(command, &context.cwd).await {
                Some(path) => path,
                None => {
                    return ToolResult::failure(
                        "SAFE_EXECUTABLE_NOT_FOUND",
                        format!("无法在工作区外的可信 PATH 中找到命令“{command}”。"),
                        context.started_at,
                        json!({"command":command}),
                    );
                }
            }
        };
        let timeout_ms = input
            .get("timeoutMs")
            .and_then(Value::as_u64)
            .unwrap_or(DEFAULT_TIMEOUT_MS);
        let mut process = Command::new(executable);
        process
            .args(&args)
            .current_dir(&cwd)
            .stdin(Stdio::null())
            .stdout(Stdio::piped())
            .stderr(Stdio::piped())
            .kill_on_drop(true)
            .env_remove("GIT_EXTERNAL_DIFF")
            .env_remove("GIT_CONFIG")
            .env_remove("GIT_CONFIG_GLOBAL")
            .env_remove("GIT_CONFIG_SYSTEM")
            .envs(&env_overrides);
        let mut child = match process.spawn() {
            Ok(child) => child,
            Err(error) => {
                return ToolResult::failure(
                    "COMMAND_SPAWN_ERROR",
                    error.to_string(),
                    context.started_at,
                    json!({"command":command}),
                );
            }
        };
        let stdout_task = child
            .stdout
            .take()
            .map(|stdout| tokio::spawn(read_limited(stdout)));
        let stderr_task = child
            .stderr
            .take()
            .map(|stderr| tokio::spawn(read_limited(stderr)));
        enum Outcome {
            Status(std::process::ExitStatus),
            Timeout,
            Aborted,
            WaitError(String),
        }
        let outcome = tokio::select! {
            _ = context.cancellation.cancelled() => {
                let _ = child.kill().await;
                let _ = child.wait().await;
                Outcome::Aborted
            }
            _ = time::sleep(Duration::from_millis(timeout_ms)) => {
                let _ = child.kill().await;
                let _ = child.wait().await;
                Outcome::Timeout
            }
            status = child.wait() => match status {
                Ok(status) => Outcome::Status(status),
                Err(error) => Outcome::WaitError(error.to_string()),
            }
        };
        let (stdout, stdout_truncated) = match stdout_task {
            Some(task) => task.await.unwrap_or_default(),
            None => Default::default(),
        };
        let (stderr, stderr_truncated) = match stderr_task {
            Some(task) => task.await.unwrap_or_default(),
            None => Default::default(),
        };
        let truncated = stdout_truncated || stderr_truncated || matches!(outcome, Outcome::Timeout);
        let summary = stdout
            .lines()
            .chain(stderr.lines())
            .rev()
            .take(20)
            .collect::<Vec<_>>()
            .into_iter()
            .rev()
            .collect::<Vec<_>>()
            .join("\n");
        let (exit_code, signal, error) = match outcome {
            Outcome::Status(status) => {
                #[cfg(unix)]
                let signal = {
                    use std::os::unix::process::ExitStatusExt;
                    status.signal().map(|signal| signal.to_string())
                };
                #[cfg(not(unix))]
                let signal: Option<String> = None;
                let error = if status.success() {
                    None
                } else {
                    Some((
                        "NONZERO_EXIT",
                        format!("命令退出码为 {:?}", status.code()),
                        false,
                    ))
                };
                (status.code(), signal, error)
            }
            Outcome::Timeout => (
                None,
                Some("TIMEOUT".into()),
                Some((
                    "COMMAND_TIMEOUT",
                    format!("命令在 {timeout_ms}ms 后超时"),
                    true,
                )),
            ),
            Outcome::Aborted => (
                None,
                Some("ABORTED".into()),
                Some(("COMMAND_ABORTED", "命令已中断".into(), false)),
            ),
            Outcome::WaitError(message) => {
                (None, None, Some(("COMMAND_WAIT_ERROR", message, false)))
            }
        };
        let mut result = if let Some((code, message, retryable)) = error {
            let mut result = ToolResult::failure(
                code,
                message,
                context.started_at,
                json!({"command":command,"exitCode":exit_code,"signal":signal,"timeoutMs":timeout_ms}),
            );
            result.error.as_mut().unwrap().retryable = retryable;
            result.output = Some(json!({
                "exitCode":exit_code,"signal":signal,"stdout":stdout,"stderr":stderr,
                "outputSummary":summary,"truncated":truncated
            }));
            result
        } else {
            ToolResult::success(
                json!({"exitCode":exit_code,"signal":signal,"stdout":stdout,"stderr":stderr,"outputSummary":summary,"truncated":truncated}),
                context.started_at,
                json!({"command":command,"args":args,"cwd":cwd}),
            )
        };
        result.metadata =
            json!({"command":command,"args":args,"exitCode":exit_code,"signal":signal,"cwd":cwd});
        result
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn command_拒绝_shell_语法() {
        assert!(valid_command("git"));
        assert!(!valid_command("git status"));
        assert!(!valid_command("sh;rm"));
        assert!(!valid_command("/bin/ls"));
    }

    #[test]
    fn git_仅允许只读子命令() {
        assert!(check_git_args(&["status".into(), "--short".into()]).is_ok());
        assert!(check_git_args(&["commit".into()]).is_err());
        assert!(check_git_args(&["diff".into(), "--output=x".into()]).is_err());
    }
}
